### PR TITLE
Cache captured_inputs

### PR DESCRIPTION
**Pull request by jonathanchu33**

### CODE REVIEW

**Comment by jonathanchu33** [comment]
Benchmarks over 3 batches of 10 trials each

Minimum
defun_matmul_2_by_2_CPU: 156 -> 155 us
defun_matmul_2_by_2_CPU_async: 84 -> 82.5 us
defun_matmul_2_by_2_with_signature_CPU: 206.5 -> 201 us

Mean
defun_matmul_2_by_2_CPU: 177 -> 173 us
defun_matmul_2_by_2_CPU_async: 86.5 -> 85 us
defun_matmul_2_by_2_with_signature_CPU: 223 -> 233 us

Median
defun_matmul_2_by_2_CPU: 176 -> 173 us
defun_matmul_2_by_2_CPU_async: 85.5 -> 83.5 us
defun_matmul_2_by_2_with_signature_CPU: 224 -> 224 us

File: tensorflow/python/eager/function.py [hunk 1]
```diff
@@ -2033,9 +2037,7 @@ def captured_inputs(self):
 
     self.__call__(*args) passes `args + self.captured_inputs` to the function.
```
Comment (kkimdev) [inline tensorflow/python/eager/function.py:2038]: Update comment.

File: tensorflow/python/eager/function.py [hunk 2]
```diff
@@ -1535,6 +1535,10 @@ def __init__(self,
               inputs=self.inputs,
               captured=self._captured_inputs,
               closures=self._captured_closures))
+
+    self._captured_inputs += nest.flatten(
```
Comment (kkimdev) [inline tensorflow/python/eager/function.py:1539]: Does it have any implications on other places that uses `self._captured_inputs` (if there is any)?

**Review by kkimdev — CHANGES_REQUESTED** [review]

File: tensorflow/python/eager/function.py (hunk 2 above)
Comment (jonathanchu33) [reply tensorflow/python/eager/function.py:1539]: Doesn't look like it

**Comment by kkimdev** [comment]
Please take a look at these Ubuntu CPU CI fails and see if you can reproduce https://source.cloud.google.com/results/invocations/f185b429-a41f-471c-9f59-9deb5a016144/targets/%2F%2Ftensorflow%2Fpython%2Feager:function_test/tests;group=__main__.MultiDeviceTest;test=testDeferredCaptureTypeError;row=1

File: tensorflow/python/eager/function.py (hunk 2 above)
Comment (kkimdev) [reply tensorflow/python/eager/function.py:1539]: I think the failing tests, "testDeferredCapture*", in the other comment could be related,

**Review by kkimdev — CHANGES_REQUESTED** [review]

File: tensorflow/python/eager/function.py (hunk 2 above)
Comment (sun51) [inline tensorflow/python/eager/function.py:1539]: this is already defined in line 1499. you need to choose a different name.

**Comment by jonathanchu33** [comment]
Closing this PR because updates to the return values of the function calls `x()` cannot be appropriately cached. `testDeferredCapture` is an example

File: tensorflow/python/eager/function.py (hunk 2 above)
Comment (jonathanchu33) [reply tensorflow/python/eager/function.py:1539]: Confirmed, it turns out that a certain return value cannot be cached. See comment at end of discussion

**Pull request closed without merging**